lib/improver/nbhood.py

```diff
@@ -0,0 +1,161 @@
+
+    The kernel will presently only work with projections in which the
+    x grid point spacing and y grid point spacing are constant over the
+    entire domain, such as the UK national grid projection.
+
+    A maximum kernel radius of 500 grid cells is imposed in order to
+    avoid computational ineffiency and possible memory errors.
+
+    """
+
+    # Max extent of kernel in grid cells.
+    MAX_KERNEL_CELL_RADIUS = 500
+
+    def __init__(self, radius_in_km, unweighted_mode=False):
+        """
+        Create a neighbourhood processing plugin that applies a smoothing
+        kernel to points in a cube.
+
+        Parameters
+        ----------
+
+        radius_in_km : float
+            The radius in kilometres of the neighbourhood kernel to
+            apply. Rounded up to convert into integer number of grid
+            points east and north, based on the characteristic spacing
+            at the zero indices of the cube projection-x/y coords.
+
+        unweighted_mode : boolean
+            If True, use a circle with constant weighting.
+            If False, use a circle for neighbourhood kernel with
+            weighting decreasing with radius.
+
+        """
+        self.radius_in_km = float(radius_in_km)
+        self.unweighted_mode = bool(unweighted_mode)
+
+    def __str__(self):
+        result = ('<NeighbourhoodProcessing: radius_in_km: {};' +
+                  'unweighted_mode: {}>')
+        return result.format(
+            self.radius_in_km, self.unweighted_mode)
+
+    def get_grid_x_y_kernel_ranges(self, cube):
+        """Return grid cell numbers east and north for the kernel."""
+        try:
+            x_coord = cube.coord("projection_x_coordinate").copy()
+            y_coord = cube.coord("projection_y_coordinate").copy()
+        except iris.exceptions.CoordinateNotFoundError:
+            raise ValueError("Invalid grid: projection_x/y coords required")
+        x_coord.convert_units("metres")
+        y_coord.convert_units("metres")
+        d_north_metres = y_coord.points[1] - y_coord.points[0]
+        d_east_metres = x_coord.points[1] - x_coord.points[0]
+        grid_cells_y = int(self.radius_in_km * 1000 / abs(d_north_metres))
+        grid_cells_x = int(self.radius_in_km * 1000 / abs(d_east_metres))
+        if grid_cells_x == 0 or grid_cells_y == 0:
+            raise ValueError(
+                ("Neighbourhood processing radius of " +
+                 "{0} km ".format(self.radius_in_km) +
+                 "gives zero cell extent")
+            )
+        if (grid_cells_x > self.MAX_KERNEL_CELL_RADIUS or
+                grid_cells_y > self.MAX_KERNEL_CELL_RADIUS):
+            raise ValueError(
+                ("Neighbourhood processing radius of " +
+                 "{0} km ".format(self.radius_in_km) +
+                 "exceeds maximum grid cell extent")
+            )
+        return grid_cells_x, grid_cells_y
+
+    def process(self, cube):
```

Comment:
I think we should aim to just always either accept a cube OR accept a cubelist. In the future, as we discussed, we could have magic decorators to handle it and type hints in Python 3 to reject invalid stuff.